File: tools/tcam-capture/tcam_capture/DeviceDialog.py
# Copyright 2017 The Imaging Source Europe GmbH
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from .TcamDevice import TcamDevice, TcamDeviceIndex
from .ImageProvider import ImageProvider

from PyQt5.QtWidgets import (QDialog, QGridLayout, QVBoxLayout,
                             QLabel, QLineEdit, QComboBox,
                             QDialogButtonBox, QListWidget, QListWidgetItem,
                             QPushButton, QHBoxLayout, QWidget)
from PyQt5.QtGui import QPixmap, QIcon

from PyQt5.QtCore import Qt, QSize

import logging

log = logging.getLogger(__file__)


class SelectionWidget(QListWidgetItem):

    def __init__(self, device: TcamDevice, parent=None):
        super(SelectionWidget, self).__init__(parent)
        self.device = device
        self.image_provider = ImageProvider()
        self.setup_ui()

    def setup_ui(self):

        self.layout = QVBoxLayout()
        self.setTextAlignment(Qt.AlignHCenter)
        self.setText(self.device.model + "\n" + self.device.serial)
        self.select_icon()

    def select_icon(self):

        self.setIcon(self.image_provider.get_default())


class DeviceDialog(QDialog):

    def __init__(self, indexer: TcamDeviceIndex, parent=None):
        super(DeviceDialog, self).__init__(parent)

        self.indexer = indexer
        self.indexer.update_device_list.connect(self.update_device_list)

        self.setWindowTitle("Tcam-Capture Device Selection")
        self.setMinimumSize(640, 480)
        self.layout = QVBoxLayout(self)

        self.selected_item = None

        self.test_label = QLabel("Please select the device that shall be used:")

        self.layout.addWidget(self.test_label)

        self.camera_box = QListWidget(self)
        self.camera_box.setFlow(QListWidget.LeftToRight)
        self.camera_box.setGridSize(QSize(180, 180))
        self.camera_box.setViewMode(QListWidget.IconMode)
        self.camera_box.setResizeMode(QListWidget.Adjust)
        self.camera_box.setMovement(QListWidget.Static)
        self.camera_box.setWordWrap(True)
        self.camera_box.setAutoFillBackground(True)

        self.camera_box.setDragEnabled(False)
        self.camera_box.setTabKeyNavigation(True)

        self.camera_box.setIconSize(QSize(100, 100))

        # handle double click
        self.camera_box.itemActivated.connect(self.item_double_clicked)
        self.camera_box.itemDoubleClicked.connect(self.item_double_clicked)
        # handle a single click
        self.camera_box.itemChanged.connect(self.item_selected)
        self.camera_box.itemClicked.connect(self.item_selected)

        self.layout.addWidget(self.camera_box)

        # OK and Cancel buttons
        self.buttons = QDialogButtonBox(
            QDialogButtonBox.Ok | QDialogButtonBox.Cancel,
            Qt.Horizontal, self)
        self.layout.addWidget(self.buttons)

        self.buttons.accepted.connect(self.accept)
        self.buttons.rejected.connect(self.reject)

        self.setLayout(self.layout)

        # ensure box is populated when we start
        dev_list = self.indexer.device_list
        if dev_list:
            self.update_device_list(dev_list)

    def item_selected(self, item):
        """
        SLOT for self.camera_box.itemClicked and
        self.camera_box.itemChanged
        """
        self.selected_item = item

    def item_double_clicked(self, item):
        """SLOT for self.camera_box.itemDoubleClicked"""

        self.item_selected(item)

        self.accept()

    def get_selected_device(self):
        """
        Returns the curreclty selected TcamDevice may return None
        """

        if self.selected_item:
            return self.selected_item.device
        return None

    def update_device_list(self, device_list):
        """
        SLOT for indexer
        """

        # remember selected dev so that we can
        # keep the highlight when the list changes
        if self.selected_item:
            selected_dev = self.selected_item.device
        else:
            selected_dev = None

        self.camera_box.clear()

        entries = []
        for dev in device_list:
            item = SelectionWidget(dev)
            entries.append(item)
            self.camera_box.addItem(item)
            if selected_dev and dev.serial == selected_dev.serial:
                item.setSelected(True)

    @staticmethod
    def get_device(indexer, parent=None):
        dialog = DeviceDialog(indexer)

        result = dialog.exec_()

        dialog.hide()
        selected_device = dialog.get_selected_device()

        return (result == QDialog.Accepted, selected_device)

File: tools/tcam-capture/devicedialog.cpp
/*
 * Copyright 2021 The Imaging Source Europe GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "devicedialog.h"

#include "devicewidget.h"

#include "ui_devicedialog.h"

DeviceDialog::DeviceDialog(std::shared_ptr<Indexer> index, QWidget* parent)
    : QDialog(parent), ui(new Ui::DeviceDialog), _index(index)
{
    ui->setupUi(this);

    // prevent dialog maximization 
    setWindowFlags(windowFlags() | Qt::Tool);

    QObject::connect(_index.get(), &Indexer::new_list, this, &DeviceDialog::update_device_listing);
    QObject::connect(_index.get(), &Indexer::new_device, this, &DeviceDialog::new_device);
    QObject::connect(_index.get(), &Indexer::device_lost, this, &DeviceDialog::lost_device);

    m_device_list = _index->get_device_list();

    fill_list();
}

DeviceDialog::~DeviceDialog()
{
    delete ui;
}

void DeviceDialog::fill_list()
{
    auto selection = this->ui->listWidget->currentItem();
    Device selected_dev;
    bool have_device = false;
    if (selection)
    {
        have_device = true;
        selected_dev = ((const DeviceWidget*)selection)->get_device();
    }

    this->ui->listWidget->blockSignals(true);
    this->ui->listWidget->clear();

    for (const auto& dev : m_device_list)
    {
        auto w = new DeviceWidget(dev);
        this->ui->listWidget->addItem(w);
        if (have_device && dev == selected_dev)
        {
            w->setSelected(true);
        }
    }

    this->ui->listWidget->update();
    this->ui->listWidget->blockSignals(false);
}

Device DeviceDialog::get_selected_device() const
{
    return m_selected_device;
}


void DeviceDialog::set_format_handling(FormatHandling handling)
{
    ui->buttonBox->blockSignals(true);
    if (handling == FormatHandling::Auto)
    {
        ui->button_auto->setChecked(true);
    }
    else
    {
        ui->button_dialog->setChecked(true);
    }
    ui->buttonBox->blockSignals(false);
}


FormatHandling DeviceDialog::get_format_handling() const
{
    if (ui->button_auto->isChecked())
    {
        return FormatHandling::Auto;
    }
    return FormatHandling::Dialog;
}

void DeviceDialog::on_buttonBox_accepted()
{
    hide();
}

void DeviceDialog::new_device(const Device& new_device)
{
    auto w = new DeviceWidget(new_device);
    this->ui->listWidget->blockSignals(true);
    this->ui->listWidget->addItem(w);
    this->ui->listWidget->blockSignals(false);
}

void DeviceDialog::lost_device(const Device& lost_device)
{
    for (int i = 0; i < this->ui->listWidget->count(); ++i)
    {
        auto itemf = static_cast<DeviceWidget*>(this->ui->listWidget->item(i));
        if (*itemf == lost_device)
        {
            QListWidgetItem* item = this->ui->listWidget->takeItem(i);

            delete item;
            break;
        }
    }
}

void DeviceDialog::update_device_listing(const std::vector<Device>& /*dev_list*/)
{}

void DeviceDialog::on_listWidget_currentItemChanged(QListWidgetItem* current,
                                                    QListWidgetItem* /*previous*/)
{
    if (current)
    {
        m_selected_device = ((DeviceWidget*)current)->get_device();
    }

}

void DeviceDialog::on_listWidget_itemDoubleClicked(QListWidgetItem* item)
{
    if (item)
    {
        m_selected_device = ((DeviceWidget*)item)->get_device();
    }
    this->accept();
}

void DeviceDialog::on_listWidget_itemClicked(QListWidgetItem* /*item*/)
{
    m_device_is_selected = false;

}
